File: javascript/backend.js
////////////////////////////////////////////////////////////////////////////////
//
// Supabase Lab
//
// backend.py - Supabase-oriented functions
//
// Copyright (c) 2024, 2025 Andy Warmack
// This file is part of Supabase Lab, licensed under the MIT License.
// See the LICENSE file in the project root for more information.
////////////////////////////////////////////////////////////////////////////////

const axios = require('axios');

const { createClient } = require('@supabase/supabase-js');

const core = require('./core');

////////////////////////////////////////////////////////////////////////////////
// connect to Supabase
////////////////////////////////////////////////////////////////////////////////

const connect = async () => {
  // initialize the Supabase client
  let url = core.Session.url;
  let api_key = null;
  if ('url' in core.Session.config) {
    url = core.Session.config['url'];
  }
  if ('api_key' in core.Session.config) {
    api_key = core.Session.config['api_key'];
  }
  core.writeln(`Connecting to ${url}`);
  if (api_key == null) {
    core.supabase_error_print('No api_key configuration found.');
    return;
  }
  core.Session.config['url'] = url;
  core.Session.config['api_key'] = api_key;
  try {
    core.Session.supabase = createClient(url, api_key);
    core.writeln('Ready to login.');
  } catch (error) {
    core.handle_error('connect', error, true);
  }
};

////////////////////////////////////////////////////////////////////////////////
// run a Supabase Edge Function
////////////////////////////////////////////////////////////////////////////////

const edge_function = async (url, payload) => {
  let headers = {
    "Authorization": `Bearer ${core.Session.jwt_token}`,
    "Content-Type": "application/json"
  };
  axios.post(url, payload, { headers: headers, timeout: 5000 })
    .then(response => {
      if (core.verbose()) {
        show_response(response);
      } else {
        core.writeln(response.data);
      }
    })
    .catch(error => {
      core.supabase_error_print(`Axios error: ${error.message}`);
      if (error.code) {
        core.supabase_error_print(`Axios error code: ${error.code}`);
      }
    });
};

////////////////////////////////////////////////////////////////////////////////
// subscribe to a channel
////////////////////////////////////////////////////////////////////////////////

async function subscribe_channel(channel_name) {
  const subscription = core.Session.supabase.channel(channel_name);
  subscription
    .subscribe((status) => {
      if (status === 'SUBSCRIBED') {
        core.writeln(`Subscribed channel: ${channel_name}`);
      } else if (status === 'CLOSED') {
        core.writeln(`Closed channel: ${channel_name}`);
      } else if (status === 'RECONNECTING') {
        core.writeln(`Reconnecting to channel: ${channel_name}`);
      } else if (status === 'ERROR') {
        core.supabase_error_print(`Error subscribing to channel: ${channel_name}`);
      }
    });
  register_channel(channel_name, subscription);
  return subscription;
}

////////////////////////////////////////////////////////////////////////////////
// unsubscribe from a channel
////////////////////////////////////////////////////////////////////////////////

async function unsubscribe_channel(channel_name) {
  const subscription = core.Session.subscriptions[channel_name];
  if (subscription) {
    const { error: unsubscribeError } = await subscription.unsubscribe();
    if (unsubscribeError) {
      core.supabase_error_print(unsubscribeError);
    } else {
      delete core.Session.subscriptions[channel_name];
    }
  }
}

////////////////////////////////////////////////////////////////////////////////
// register a channel for later reference
////////////////////////////////////////////////////////////////////////////////

function register_channel(channel_name, subscription) {
  if (subscription) {
    core.Session.subscriptions[channel_name] = subscription;
  } else {
    core.supabase_error_print(`${channel_name}: Invalid subscription object.`);
  }
}

////////////////////////////////////////////////////////////////////////////////
// list registered channels
////////////////////////////////////////////////////////////////////////////////

async function list_channels() {
  for (let channel_name in core.Session.subscriptions) {
    core.writeln(channel_name);
  }
}

////////////////////////////////////////////////////////////////////////////////
// listen for broadcast messages on a channel
////////////////////////////////////////////////////////////////////////////////

async function listen_to_broadcast_channel(channel_name, event) {
  const subscription = await subscribe_channel(channel_name);
  subscription
    .on('broadcast', { event: event }, (payload) => {
      core.writeln(`FROM ${channel_name}: ${JSON.stringify(payload)}`);
    });
}

////////////////////////////////////////////////////////////////////////////////
// send broadcast message on a channel
////////////////////////////////////////////////////////////////////////////////

async function send_to_broadcast_channel(channel_name, event, message_text) {
  if (message_text == '' || !message_text) {
    core.error_print('Empty message not sent.');
    return;
  }
  const message = {
    type: 'broadcast',
    event: event,
    payload: {
      message: message_text,
    }
  };
  core.Session.supabase.channel(channel_name).send(message);
  core.writeln(`TO ${channel_name}: ${JSON.stringify(message)}`);
}

////////////////////////////////////////////////////////////////////////////////
// listen for presence signals on a channel
////////////////////////////////////////////////////////////////////////////////

async function sync_track_presence(channel_name) {
  let channel = core.Session.subscriptions[channel_name];
  if (!channel) {
    channel = await subscribe_channel(channel_name);
  }
  channel
    .on('presence', { event: 'sync' }, () => {
      const newState = channel.presenceState();
      core.writeln(`sync ${JSON.stringify(newState, null, 2)}`);
    })
    .on('presence', { event: 'join' }, ({ key, newPresences }) => {
      core.writeln(`join ${key} ${JSON.stringify(newPresences, null, 2)}`);
    })
    .on('presence', { event: 'leave' }, ({ key, leftPresences }) => {
      core.writeln(`leave ${key} ${JSON.stringify(leftPresences, null, 2)}`);
    });
}

////////////////////////////////////////////////////////////////////////////////
// send presence state on a channel
////////////////////////////////////////////////////////////////////////////////

async function send_presence(channel_name) {
  let channel = core.Session.subscriptions[channel_name];
  if (!channel) {
    channel = await subscribe_channel(channel_name);
  }
  const userStatus = {
    user: core.Session.config['email'],
    online_at: new Date().toISOString(),
  };
  const presenceTrackStatus = await channel.track(userStatus);
  core.writeln(JSON.stringify(presenceTrackStatus, null, 2));
}

////////////////////////////////////////////////////////////////////////////////
// stop presence tracking on a channel
////////////////////////////////////////////////////////////////////////////////

async function stop_presence(channel_name) {
  const channel = core.Session.subscriptions[channel_name];
  if (channel) {
    channel.untrack();
  }
}

////////////////////////////////////////////////////////////////////////////////
// listen for database table changes
////////////////////////////////////////////////////////////////////////////////

const listen_to_table = async (table_name) => {
  core.Session.supabase
    .channel('table_changes')
    .on(
      'postgres_changes',
      { event: '*', schema: 'public', table: table_name },
      (payload) => {
        core.writeln(payload);
      }
    )
    .subscribe();
  core.writeln(`Listening for changes on table: ${table_name}`);
};

////////////////////////////////////////////////////////////////////////////////
// display response values after sending a request to Supabase
////////////////////////////////////////////////////////////////////////////////

function show_response(response) {
  core.writeln(`Response URL: ${response.config.url}`);
  core.writeln(`Status Code: ${response.status}`);
  core.writeln(`Status Text: ${response.statusText}`);
  core.writeln(`Request Method: ${response.request.method}`);
  core.writeln('Response Data:');
  for (let key in response.data) {
    core.writeln(`   ${key}: ${response.data[key]}`);
  }  
  core.writeln('Response Headers:');
  for (let key in response.headers) {
    core.writeln(`   ${key}: ${response.headers[key]}`);
  }  
}

////////////////////////////////////////////////////////////////////////////////
// sign in the user using email/password authentication
////////////////////////////////////////////////////////////////////////////////

const sign_in = async (email, password) => {
  if (email == '' || password == '') {
    core.writeln('Invalid email or password.');
    return;
  }
  core.writeln('Logging in...');

  try {
    const { data, error } = await core.Session.supabase.auth.signInWithPassword({ email: email, password: password });
    if (error) {
      core.supabase_error_print(`Login failed: ${error.message}`);
    } else {
      core.Session.jwt_token = data.session.access_token;
      let session_user_email = data.session.user.email;
      let last_sign_in_at = core.show_time(data.user.last_sign_in_at);
      core.Session.authenticated = true;
      core.update_prompt();
      core.writeln(`${session_user_email} logged in.`);
      core.writeln(`Last login: ${last_sign_in_at}`);
    }
  } catch (error) {
    if (error instanceof TypeError) {
      core.supabase_error_print(error.message);
    } else {
      core.supabase_error_print('Unexpected error:', error);
    }
  }
};

////////////////////////////////////////////////////////////////////////////////
// sign out
////////////////////////////////////////////////////////////////////////////////

const sign_out = async () => {
  core.writeln('Logging out.');
  core.Session.supabase.auth.signOut();
  core.Session.jwt_token = null;
  core.Session.authenticated = false;
};

module.exports.connect = connect;
module.exports.edge_function = edge_function;
module.exports.subscribe_channel = subscribe_channel;
module.exports.unsubscribe_channel = unsubscribe_channel;
module.exports.register_channel = register_channel;
module.exports.list_channels = list_channels;
module.exports.listen_to_broadcast_channel = listen_to_broadcast_channel;
module.exports.send_to_broadcast_channel = send_to_broadcast_channel;
module.exports.sync_track_presence = sync_track_presence;
module.exports.send_presence = send_presence;
module.exports.stop_presence = stop_presence;
module.exports.listen_to_table = listen_to_table;
module.exports.sign_in = sign_in;
module.exports.sign_out = sign_out;

File: python/backend.py
################################################################################
#
# Supabase Lab
#
# backend.py - Supabase-oriented functions
#
# Copyright (c) 2024, 2025 Andy Warmack
# This file is part of Supabase Lab, licensed under the MIT License.
# See the LICENSE file in the project root for more information.
################################################################################

import asyncio
import json
from datetime import datetime, timezone
try:
    from realtime._async.client import AsyncRealtimeClient
    from realtime.types import RealtimeSubscribeStates
    REALTIME_IMPORTED = True
except ImportError:
    REALTIME_IMPORTED = False
try:
    import requests
    REQUESTS_IMPORTED = True
except ImportError:
    REQUESTS_IMPORTED = False
try:
    from supabase import acreate_client
    SUPABASE_IMPORTED = True
except ImportError:
    SUPABASE_IMPORTED = False
from typing import Optional

import core

################################################################################
# check for required imports
################################################################################

def realtime_imported():
    return REALTIME_IMPORTED

def requests_imported():
    return REQUESTS_IMPORTED

def supabase_imported():
    return SUPABASE_IMPORTED

################################################################################
# connect to Supabase
################################################################################

async def connect():
    # initialize the Supabase client
    url = core.Session.url
    api_key = None
    if 'url' in core.Session.config:
        url = core.Session.config['url']
    if 'api_key' in core.Session.config:
        api_key = core.Session.config['api_key']
    print(f"Connecting to {url}")
    if api_key is None:
        core.supabase_error_print('No api_key configuration found.')
        return
    core.Session.config['url'] = url
    core.Session.config['api_key'] = api_key
    try:
        core.Session.supabase = await acreate_client(url, api_key)
        core.Session.realtime = AsyncRealtimeClient(websocket_url(), api_key)
        print('Ready to login.')
    except Exception as e: # pylint: disable=broad-exception-caught
        core.handle_error('connect()', e)

def websocket_url():
    return core.Session.config['url'].replace('http://', 'wss://').replace('https://', 'wss://') + '/realtime/v1'

def check_connection():
    if core.Session.supabase:
        return True
    core.supabase_error_print('Supabase client not connected.')
    return False

################################################################################
# run a Supabase Edge Function
################################################################################

def edge_function(url, payload):
    if not requests_imported():
        core.info_print('The Requests library is required: https://requests.readthedocs.io/')
        return
    headers = {
        "Authorization": f"Bearer {core.Session.jwt_token}",
        "Content-Type": "application/json"
    }
    try:
        response = requests.post(url, json=payload, headers=headers, timeout=10)
        response.raise_for_status()
        if core.verbose() or not response.ok:
            show_response(response)
        elif 'application/json' in response.headers.get('Content-Type', ''):
            print(response.json())
        else:
            print(response.headers.get('Content-Type', ''))
            print(response.text)
    except requests.exceptions.HTTPError as e:
        core.handle_error('edge_function()', e)
    except requests.exceptions.RequestException as e:
        core.handle_error('edge_function()', e)
    except Exception as e: # pylint: disable=broad-exception-caught
        core.handle_error('edge_function()', e, True)

################################################################################
# display subscribe state messages
################################################################################

def on_subscribe_for_channel(channel_name, for_table = False):
    def on_subscribe(status: RealtimeSubscribeStates, error: Optional[Exception]):
        if status == RealtimeSubscribeStates.SUBSCRIBED:
            if not for_table:
                print(f"Subscribed channel: {channel_name}")
            else:
                print(f"Listening for changes on table: {channel_name}")
        elif status == RealtimeSubscribeStates.CLOSED:
            print(f"Closed channel: {channel_name}")
        elif status == RealtimeSubscribeStates.TIMED_OUT:
            core.supabase_error_print('Channel connection timed out')
        elif status == RealtimeSubscribeStates.CHANNEL_ERROR:
            core.supabase_error_print(f"Error subscribing to channel: {error.message}")

    return on_subscribe

################################################################################
# subscribe to a channel
################################################################################

async def subscribe_channel(channel_name, event = 'test', force = True):
    if not check_connection():
        return False
    if channel_name in core.Session.subscriptions:
        subscription = core.Session.subscriptions[channel_name]
        return subscription
    if not force:
        core.error_print(f"Please first subscribe to channel: {channel_name}")
        return False
    client = core.Session.realtime
    await client.connect()
    channel = client.channel(channel_name, { 'type': 'broadcast', 'event': event, "config": {"broadcast": {"self": True}}})
    subscription = await channel.subscribe(on_subscribe_for_channel(channel_name))
    register_channel(channel_name, subscription)
    return subscription

################################################################################
# unsubscribe from a channel
################################################################################

async def unsubscribe_channel(channel_name):
    if not check_connection():
        return
    if channel_name not in core.Session.subscriptions:
        return
    subscription = core.Session.subscriptions[channel_name]
    if subscription:
        await subscription.unsubscribe()
        del core.Session.subscriptions[channel_name]
        print(f"Closed channel: {channel_name}")

################################################################################
# register a channel for later reference
################################################################################

def register_channel(channel_name, subscription):
    if subscription:
        core.Session.subscriptions[channel_name] = subscription
    else:
        core.supabase_error_print(f"{channel_name}: Invalid subscription object.")

################################################################################
# list registered channels
################################################################################

async def list_channels():
    for channel_name in core.Session.subscriptions:
        print(channel_name)

################################################################################
# listen for broadcast messages on a channel
################################################################################

async def listen_to_broadcast_channel(channel_name, event):
    if not check_connection():
        return
    if channel_name == '':
        core.error_print('Please specify a channel.')
        return

    await core.Session.realtime.connect()
    subscription = await subscribe_channel(channel_name)
    subscription.on_broadcast(event, lambda payload: print(f"FROM {channel_name}: {payload}"))

################################################################################
# send broadcast message on a channel
################################################################################

async def send_to_broadcast_channel(channel_name, event, payload):
    if not check_connection():
        return
    await core.Session.realtime.connect()
    subscription = await subscribe_channel(channel_name, event, False)
    if subscription is not False:
        await subscription.send_broadcast(event, payload)
        print(f"TO {channel_name}: {payload}")

################################################################################
# listen for presence signals on a channel
################################################################################

async def sync_track_presence(channel_name):
    channel = core.Session.supabase.channel(channel_name)
    register_channel(channel_name, channel)

    def on_sync():
        new_state = channel.presence_state()
        print(f"sync {json.dumps(new_state, indent=2)}")

    def on_join(key, _, new_presences):
        print(f"join {key} {json.dumps(new_presences, indent=2)}")

    def on_leave(key, _, left_presences):
        print(f"leave {key} {json.dumps(left_presences, indent=2)}")

    await channel.on_presence_sync(on_sync).on_presence_join(on_join).on_presence_leave(on_leave).subscribe()

################################################################################
# send presence state on a channel
################################################################################

async def send_presence(channel_name):
    channel = await subscribe_channel(channel_name)
    await asyncio.sleep(0.1)
    user_status = {
        "user": core.Session.config['email'],
        "online_at": datetime.now(timezone.utc).isoformat()
    }
    presence_track_status = await channel.track(user_status)
    print(f"{'ok' if presence_track_status is None else presence_track_status}")

################################################################################
# stop presence tracking on a channel
################################################################################

async def stop_presence(channel_name):
    subscription = await subscribe_channel(channel_name)
    await subscription.untrack()

################################################################################
# listen for database table changes
################################################################################

async def listen_to_table(table_name):
    client = core.Session.realtime
    await client.connect()
    channel = client.channel('table_changes', {
        'type': 'postgres_changes',
        'table': table_name,
        'event': '*'
    })
    channel.on_postgres_changes(
        "*",
        schema="public",
        callback=print
    )
    await channel.subscribe(on_subscribe_for_channel(table_name, True))
    await client.listen()

################################################################################
# display response values after sending a request to Supabase
################################################################################

def show_response(response):
    print(f"Response URL: {response.url}")
    print(f"Status Code: {response.status_code}")
    print(f"Response OK: {response.ok}")
    print(f"Elapsed Time: {response.elapsed.total_seconds()} seconds ({response.elapsed})")
    print(f"Content-Type: {response.headers.get('Content-Type', '')}")
    print(f"Cookies: {response.cookies.get_dict()}")
    print(f"Request Method: {response.request.method}")
    print(f"Request URL: {response.request.url}")
    print('Request Headers:')
    for header, value in response.request.headers.items():
        print(f"   {header}: {value}")
    print(f"Request Body: {response.request.body.decode('utf-8')}")
    print(f"History: {response.history}")
    print(f"Response JSON: {response.json()}")
    print(f"Response Test: {response.text}")

################################################################################
# sign in the user using email/password authentication
################################################################################

async def sign_in(email, password):
    if email == '' or password == '':
        print('Invalid email or password.')
        return
    print('Logging in...')
    response = None
    try:
        response = await core.Session.supabase.auth.sign_in_with_password({"email": email, "password": password})
    except Exception as e: # pylint: disable=broad-exception-caught
        if response is not None:
            core.supabase_error_print(response)
        else:
            core.supabase_error_print('No response received.')
        core.supabase_error_print(f"sign_in({email}, {'*' * min(len(password), 8)}): {e}")
        core.Session.authenticated = False

    # get JWT token and greet user
    if response is not None:
        core.Session.jwt_token = response.session.access_token
        email = response.session.user.email
        last_sign_in_at = core.show_time(response.user.last_sign_in_at)
        print(f"{email} logged in.")
        print(f"Last login: {last_sign_in_at}")
        core.Session.authenticated = True

################################################################################
# sign out
################################################################################

async def sign_out():
    print('Logging out.')
    await core.Session.supabase.auth.sign_out({'scope': 'local'})
    core.Session.jwt_token = None
    core.Session.authenticated = False
